[3d] Large scene support: shift scene's origin if the camera got too far

wonder-sk opened the pull request:
This "origin rebasing" or "origin shifting" technique is commonly used in various 3D engines when dealing with large scenes (where objects need double precision coordinates). It should greatly improve our capability to handle large scenes (e.g. more than 50km across)

For now, I have used a simple rule that when camera gets 10 km away from the world's origin, the origin is shifted (and camera is placed at the origin). The origin shift has a bit of a cost (we're updating 4x4 transform matrices of all active nodes), so we prefer not to do it on every camera change, but only when necessary. That said, I have not seen slowdowns when origin gets shifted.

Also add a checkbox to the debug panel to allow devs temporarily turn it off.

A related philosophical question from Futurama:

![image](https://github.com/user-attachments/assets/333a3f70-91c4-4eeb-8dec-25d0124570cf)


Before:

https://github.com/user-attachments/assets/ec4fea53-08a7-490b-8993-52769320e345


After:


https://github.com/user-attachments/assets/2f030fd1-ad64-4316-b349-32cebe2f598d

wonder-sk committed:
Large scene support: shift scene's origin if the camera got too far

This should greatly improve our capability to handle large scenes (e.g. more than 50km across)

Also add a checkbox to the debug panel to allow devs temporarily turn it off.

uclaros reviewed (commented):
So nice!
All the underlying work finally getting enabled!

wonder-sk committed:
Review from Stefanos + test fix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wonder-sk committed:
Fix reset view functionality

Relax the assumption that origin is always in the center of the scene

wonder-sk committed:
Fix tests - call setExtent(), not setOrigin()

wonder-sk committed:
setting origin is not required (it is done in setExtent() anyway)

wonder-sk committed:
Fix 3d map scene test

Origin may get updated in Qgs3DMapScene constructor, so let's set
ours once the Qgs3DMapScene is constructed

nyalldawson commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -315,6 +315,20 @@ void Qgs3DMapScene::onCameraChanged()
+  // with large coordinates in 32-bit floats (and if we do have large coordinates,
+  // because the scene is far from the camera, we don't care, because those errors
+  // end up being tiny when viewed from far away).
+  const float originShiftThreshold = 10'000;
+  if ( mSceneOriginShiftEnabled && mEngine->camera()->position().length() > originShiftThreshold )
```
```suggestion
  constexpr float ORIGIN_SHIFT_THRESHOLD = 10'000;
  if ( mSceneOriginShiftEnabled && mEngine->camera()->position().length() > ORIGIN_SHIFT_THRESHOLD )
```

nyalldawson committed:
Update src/3d/qgs3dmapscene.cpp

nyalldawson commented:
Looks fantastic -- great work @wonder-sk !!
